### PR TITLE
Tobias2 to dev

### PR DESCRIPTION
**Pull request by tobyy22**

### CODE REVIEW

**Comment by tobyy22** [comment]
The type conversions work but are not very nice, I will refactor them later or we can discuss it.

**Comment by tobyy22** [comment]
@YoungMasterGandalf

File: dbhydra/dbhydra_core.py [hunk 1]
```diff
@@ -17,6 +17,28 @@
...
+    "timestamp": "timestamp"
+}
+
+PYTHON_TO_MYSQL_DATA_MAPPING = {
+    'int': "int",
+    'float': "double",
+    'str': "varchar(255)",
+    'bool': "tinyint",
+    'datetime': "datetime"
+}
```
Comment (DovaX) [inline dbhydra/dbhydra_core.py:41]: nvarchar preferred + introduction of new pattern -> needs to be checked

File: dbhydra/dbhydra_core.py [hunk 2]
```diff
@@ -80,9 +102,14 @@ def new_function(instance, *args, **kw):
...
 
+# class DataMigrator:
+
+
+
 class Migrator:
-    def __init__(self, db=None):
+    def __init__(self, db=None, db2=None):
         self.db = db
+        self.db2 = db2
```
Comment (DovaX) [inline dbhydra/dbhydra_core.py:112]: I don't think this is correct - db2 only needed if migrating data, not migrating structure

File: dbhydra/dbhydra_core.py [hunk 3]
```diff
@@ -167,13 +198,24 @@ def extract_columns_and_types_from_df(self, dataframe):
         columns = list(dataframe.columns)
 
         return_types = []
+
+        if columns == []:
+            return ["id"], ["int"]
+
         for column in dataframe:
```
Comment (DovaX) [inline dbhydra/dbhydra_core.py:205]: inner part of this for loop could be separated as an auxiliary function

File: dbhydra/dbhydra_core.py [hunk 4]
```diff
@@ -279,8 +321,8 @@ def close_connection(self):
         self.connection.close()
         print("DB connection closed")
 
-    def initialize_migrator(self):
-        self.migrator = Migrator(self)
+    def initialize_migrator(self, other_db=None):
+        self.migrator = Migrator(self, other_db)
```
Comment (DovaX) [inline dbhydra/dbhydra_core.py:325]: no

File: dbhydra/dbhydra_core.py [hunk 5]
```diff
@@ -533,7 +576,8 @@ def get_all_tables(self):
         return self.database.list_collection_names()
 
     def create_table(self, name):
-        return self.database[name]
+        table = self.database[name]
```
Comment (DovaX) [inline dbhydra/dbhydra_core.py:579]: self.database very unintuitive, it should probably be self.instance + it's questionable and needs to be checked

File: dbhydra/dbhydra_core.py [hunk 6]
```diff
@@ -685,13 +735,15 @@ def _adjust_df(self, df: pd.DataFrame, debug_mode=False) -> pd.DataFrame:
 
         return df_copy
 
-    def insert_from_df(self, df, batch=1, try_mode=False, debug_mode=False, adjust_df=False):
+    def insert_from_df(self, df, batch=1, try_mode=False, debug_mode=False, adjust_df=False, insert_id=False):
```
Comment (DovaX) [inline dbhydra/dbhydra_core.py:738]: check name + default id structure

File: dbhydra/dbhydra_core.py [hunk 7]
```diff
@@ -727,21 +779,50 @@ class PostgresTable(AbstractTable):
...
+
     def get_all_types(self):
+
         information_schema_table = Table(self.db1, 'INFORMATION_SCHEMA.COLUMNS', ['DATA_TYPE'], ['nvarchar(50)'])
-        query = "SELECT DATA_TYPE FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME  = '" + self.name + "'"
+        query = "SELECT DATA_TYPE,character_maximum_length FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME  = '" + self.name + "'"
         types = information_schema_table.select(query)
-        print(types)
-        return (types)
+        only_types = [x[0] for x in types]
```
Comment (DovaX) [inline dbhydra/dbhydra_core.py:814]: types, not only_types

File: dbhydra/dbhydra_core.py [hunk 8]
```diff
@@ -752,74 +833,93 @@ def init_all_columns(cls, db1, name):
         temporary_table = cls(db1, name)
         columns = temporary_table.get_all_columns()
         types = temporary_table.get_all_types()
+
+        if "id" in columns:
+            id_col_index = columns.index("id")
+            columns.pop(id_col_index)
+            columns.insert(0, "id")
+            types.pop(id_col_index)
+            types.insert(0, "int")
```
Comment (DovaX) [inline dbhydra/dbhydra_core.py:842]: refactor

File: dbhydra/dbhydra_core.py [hunk 9]
```diff
@@ -752,74 +833,93 @@ def init_all_columns(cls, db1, name):
...
                         query += self.columns[i] + ","
                 if len(rows) < len(self.columns):
                     print(len(self.columns) - len(rows), "columns were not specified")
-                query = query[:-1] + ") VALUES "
+                if query[-1] == ',':
+                    query = query[:-1]
+                    query = query + ") VALUES "
+                elif query[-1] == '(':
+                    query = query[:-1]
+                    query = query + " VALUES "
```
Comment (DovaX) [inline dbhydra/dbhydra_core.py:882]: else: missing -> invariant operations needed

Comment (DovaX) [reply dbhydra/dbhydra_core.py:882]: critical

File: dbhydra/dbhydra_core.py [hunk 10]
```diff
@@ -752,74 +833,93 @@ def init_all_columns(cls, db1, name):
...
                     query += "'" + str(rows[k][j]) + "',"
 
 
 
                 else:
                     query += str(rows[k][j]) + ","
+            if query[-1] == ",":
+                query = query[:-1]
+            elif query[-1] == '(':
+                query = query + "DEFAULT"
```
Comment (DovaX) [inline dbhydra/dbhydra_core.py:920]: else: missing - critical

File: dbhydra/dbhydra_core.py [hunk 11]
```diff
@@ -956,8 +1060,19 @@ def init_all_columns(cls, db1, name):
 
         temporary_table = cls(db1, name)
         values = temporary_table.get_columns_types()
-        columns = values[0]
-        types = values[1]
+        columns = values[0][1:]
+        types = values[1][1:]
```
Comment (DovaX) [inline dbhydra/dbhydra_core.py:1064]: omits id - check

File: dbhydra/dbhydra_core.py [hunk 12]
```diff
@@ -1002,6 +1117,8 @@ def get_columns_types(self):
         for dict_j in self.collection.find():
             self.print_nested_keys(dict_j, columns, types)
         types = self.get_all_types(types)
+        self.columns = columns
+        self.types = types
```
Comment (DovaX) [inline dbhydra/dbhydra_core.py:1121]: new pattern - check

**Review by DovaX — APPROVED** [review]

**Pull request merged**